### ISSUE TITLE
FlexScrollView integration

### ISSUE THREAD

**Issue by mcbain** [gadicc/meteor-famous-views#178]
https://github.com/IjzerenHein/famous-flex/blob/master/tutorials/FlexScrollView.md

Providing a fview plugin for complex views like FlexScrollView should follow a clear concept. IMHO most users would prefer a declarative approach, than programing javascript. So lets collect some ideas here.
## Basic options

All direct options for FlexScrollview should be reactivly e.g. by template-helpers

``` jade
+FlexScrollView flow=templatehelper
```
## Sticky Headers

![image](https://github.com/IjzerenHein/famous-flex/raw/master/tutorials/stickyheaders.gif)

``` jade
+FlexScrollView
  +famousEach items
    +famousIf isHeader
      +xyzView target='header'
    +else
      +xyzView
```

**Comment by IjzerenHein** [comment]
Gents, I've released v0.1.4 of famous-flex.
It contains several improvements, amongst which new '.swap' and '.ensureVisible' functions and 'scroll' and 'pagechange' events.
https://github.com/IjzerenHein/famous-flex/releases

As for new years resolutions, mine is to learn and get started with Meteor :)

Happy new year and let's make 2015 count! :)

**Comment by trusktr** [comment]
This is awesome! Can wait to see this with the new ScrollView when it finally comes out.

**Comment by PEM--** [comment]
Happy new year :tada: :balloon: :balloon: It's going to be an awesome year!

**Comment by mcbain** [comment]
Happy new year 2015.

Am Sonntag, 4. Januar 2015 schrieb Pierre-Eric Marchandet :

> Happy new year [image: :tada:] [image: :balloon:] [image: :balloon:] It's
> going to be an awesome year!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gadicc/meteor-famous-views/issues/178#issuecomment-68631980>
> .
>


-- 
Jens Zastrow

Geschäftsführender Gesellschafter / Managing Partner


MJ Networks GmbH
Strausberger Str. 44
D-10243 Berlin

Fon +49 30 530 86368
Fax +49 30 530 86370
jens.zastrow@mj-networks-gmbh.de <maher.chamsi@mj-networks-gmbh.de>


----------------------------------------------------------
MJ Networks GmbH
Geschäftsführer: Maher Chamsi, Jens Zastrow
Handelsregister: Amtsgericht Charlottenburg, HRB 117014B
Sitz der Gesellschaft: Berlin
----------------------------------------------------------

**Comment by gadicc** [comment]
*early* release, still a lot missing, but a good starting point I think.  very fun to play with :)

http://fview-flex.meteor.com/
https://github.com/gadicc/fview-flex/

**Comment by splendido** [comment]
:+1:

**Comment by Wenape** [comment]
It looks awesome!

**Comment by IjzerenHein** [comment]
Awesome :)))

**Comment by PEM--** [comment]
Starred :star:

**Comment by Lauricio** [comment]
:+1:

**Comment by trusktr** [comment]
:star2:

**Comment by IjzerenHein** [comment]
Awesome work, love how all the options can be configured on one line!

:+1: and starred!

**Comment by vplifer** [comment]
Cool! It's awesome!

**Comment by gadicc** [comment]
Thanks everyone!

@IjzerenHein, thanks... that's actually a mistake though hehe... you can still currently pass all those options in one line if you JSON-encode them and pass to `layoutOptions`.  It wouldn't actually be too hard to tweak the code to accept them directly, but I think maybe the separation is better.  I'll think about it.  Generally it looks like this:

```jade
  +ContainerSurface perspective=500 overflow="hidden"
    +FlexScrollView layout="WheelLayout" direction="Y" layoutOptions=layoutOptions
      +famousEach surfaces
        +Surface style=style
          | #{_id}
```

where `layoutOptions` is a reactive helper (more info in the README, but it's very meteory).  In general I think you're going to love how all the reactivity works when you get into Meteor... it's a really good fit for complex layouts that depend on multiple factors.

**Comment by gadicc** [comment]
v0.0.2 is out with famono support :)

**Comment by IjzerenHein** [comment]
Awesome, thanks for the explanation @gadicc.

Be careful with `overflow: hidden` though. I've ran into a lot of problems will that lately. Everything from performance degradation, [z-indexes not being respected anymore](https://github.com/Famous/famous/issues/493), and drawing issues on Android...

**Comment by gadicc** [comment]
Bummer :(  It's an officially sanctioned pattern, no?  I'll switch to a higher z-Index on the header instead in this case, but a ContainerSurface with overflow:hidden is so useful for things like https://atmospherejs.com/pierreeric/fview-devices, etc.  Do you happen to know if Famo.us is working on this and how it will affect mixed mode?  Either way thanks for the heads up!

**Comment by trusktr** [comment]
@gadicc @IjzerenHein Yep, this is a regression in Chrome. I remember this working, specifically that nested divs have their own z-indexes inside their parent with overflow:hidden. They borked it. xD

@gadicc Your FLexScrollView, is it using custom scroll behavior? I noticed it doesn't have the "bounce" effect ScrollViews normally have.

**Comment by jordangarside** [comment]
@trustkr FlexScrollView doesn't have bounce when you use the mousewheel, try dragging or using touch events

**Comment by gadicc** [comment]
@trusktr, I'm using [this nodeSpring default](https://github.com/gadicc/fview-flex/blob/4ca8f1788ae0998f0c9fa638d6805ac70158f76b/lib/FlexScrollView.js#L30-L33) which can be overridden with an attribute.  Yeah, no bounce with mousewheel, but you can get it with a mouse "flick" with the simulated mouseMove stuff (which I enabled by default).  It takes some practice though, and you don't get it every time... but if you keep trying you'll definitely see the effect :)  And I'm sure it works perfect on an actual touch device.  Dragging works fine, it's just the flick that's hard.

**Comment by trusktr** [comment]
Oh okay. I also noticed that with the mousewheel, the scrolling stops randomly all of a sudden without slowing down or reaching the end. Have you experienced that? I'm actually in Mac OS X using two-finger touch on the glass pad to fling (since with native scrolling in Mac OS X you 'fling' with the touch pad).

**Comment by gadicc** [comment]
@trusktr, could also be "gaps" in the surfaces not piping events back to the scrollview.  this was addressed in the fview-flex demo though; did you have this problem in the demo or in your own code?  what about the original famous-flex (upstream) demo?  and how does it compare to famo.us' native scrollview?

@jordangarside, nice site :)  it's not famo.us though, right?  looks like [malihu-custom-scrollbar-plugin](http://manos.malihu.gr/repository/custom-scrollbar/demo/examples/complete_examples.html)?

**Comment by jordangarside** [comment]
@gadicc Thanks! The site is just plain meteor and twitter bootstrap, and you're right, the scroll plugin is Malibu's. I say everything I use on the about page :). I'm working on the mobile app for it now and just using famo.us for everything. I debated using Polymer, but it makes every app look the same and doesn't allow much customization for animations (it's also fairly janky on mobile).

**Comment by trusktr** [comment]
@gadicc I was trying only the demo at http://fview-flex.meteor.com/. If I scroll slowly, it works just fine, even when the cursor is in the white spaces between the elements. The problem only happens when I try to 'fling' the view by two-finger throw on the Mac's touchpad. I modified the famous-angular scrollview demo, and made every odd item not pipe to the event handler. The behavior feels like what I'm experiencing in the fview-flex demo: http://plnkr.co/edit/COTKOxnMgL0LmawhpnUz?p=preview

**Comment by gadicc** [comment]
@trusktr, It's difficult for me to trouble shoot this since I don't have a Mac :(  Your modified scrollview does sound like the "crack" issue though, but yeah, actually that is the point of the ContainerSurface... it creates a div in the background and is auto-piped to the FlexScrollView.  I can consistently scroll all the way through, unlike your repro - even with spacing of say, 200.  I'll be very interested if you find out anything more about this.  What about the original flex demo at https://rawgit.com/IjzerenHein/famous-flex-demo/master/dist/index.html ?

**Comment by trusktr** [comment]
@gadicc Interesting, I don't seem to have the problem with the original flex demo!

**Comment by IjzerenHein** [comment]
@trusktr @gadicc If you're having scrolling issues, then most of the time this is related to events which have been piped too often. If both the events from the ContainerSurface and the surfaces inside are piped you'll typically see random stops and such.

**Comment by gadicc** [comment]
Wow, @IjzerenHein, you're totally right!  :bow: 

@trusktr, try the demo site now and if that fixes your issue I'll make a new release.  It's definitely a lot easier for me to flick with the mouse now too.

**Comment by trusktr** [comment]
I just tried http://fview-flex.meteor.com/, it's still having the same problem, but it seems to happen only if I fling it at a fast enough speed, otherwise it's smooth.

**Comment by gadicc** [comment]
@trusktr, copy that, thanks.  so an improvement but not completely solved.  and on https://rawgit.com/IjzerenHein/famous-flex-demo/master/dist/index.html fast flicking works?

**Comment by trusktr** [comment]
@gadicc Yes indeed, fast flicking is working there. I'm curious to know what's the difference!

**Comment by trusktr** [comment]
In [this example] (famous might be outdated) I'm experiencing the same behavior.

**Comment by gadicc** [comment]
Hey, your link is broken... still hoping to track this down when I have a chance.

**Comment by gadicc** [comment]
@trusktr, sorry, unfortunately i can't test this myself... but just tried something else, could you test http://fview-flex.meteor.com/ again please?

**Comment by trusktr** [comment]
Oops, fixed the link.

**Comment by gadicc** [comment]
Oh :(   Thanks, the video helps to understand a bit better.

So, the other link you gave "[Scrollview in HeaderFooterLyaout](http://jsfiddle.net/ZCMaL/21/)" is using the native famo.us Scrollview.  So maybe it's an Engine issue with Mac OS X.  But you said that in the [original famous-flex demo](https://rawgit.com/IjzerenHein/famous-flex-demo/master/dist/index.html) you don't notice this.

The fact that my most recent change made the situation worse makes me think that we might still get to fix this, but I'm at a bit of a loss.  @IjzerenHein, if you have any more insight that would be greatly appreciated.  Are you on a Mac?  trusktr's is referring solely to flicking with two fingers on the touchpad on Mac OS.

So basically, as you suggested, not piping the surfaces as well as the container make a big difference.  However, from the above feedback, `container.pipe(flexscrollview)` worked *better* than what I tried now from famous-flex source:

```js
      flexscrollview.subscribe(container);
      EventHandler.setInputHandler(container, flexscrollview);
      EventHandler.setOutputHandler(container, flexscrollview);      
```

Beyond that, I guess it could be part of the overhead of famous-views - wrapping stuff in RenderNodes and MeteorFamousViews.  But the `flexscrollview` and `container` above are the actual pre-wrapped nodes.

**Comment by trusktr** [comment]
Side topic: I noticed Famo.us guides recommend using two event handlers per view/thing/class/whatever (one for input, one for output). This is an awkward pattern IMO! I'm using only a single handler per view/thing/class/whatever in my stuff and it's working out great!
